client-gen/src/generator.rs

```diff
@@ -0,0 +1,438 @@
+        self.code.push_str(&client_gen.code);
+    }
+
+    fn visit_type(&mut self, t: &'ast Type) {
+        let mut type_gen = TypeGenerator::new(t.name());
+        type_gen.visit_type(t);
+        self.code.push_str(&type_gen.code);
+    }
+}
+
+struct ServiceGenerator {
+    service_name: String,
+    code: String,
+}
+
+impl ServiceGenerator {
+    fn new(service_name: String) -> Self {
+        Self {
+            service_name,
+            code: String::new(),
+        }
+    }
+}
+
+impl<'ast> Visitor<'ast> for ServiceGenerator {
+    fn visit_service(&mut self, service: &'ast Service) {
+        self.code
+            .push_str(&format!("pub trait {} {{\n", self.service_name));
+        visitor::accept_service(service, self);
+        self.code.push_str("}\n");
+    }
+
+    fn visit_func(&mut self, func: &'ast Func) {
+        let mutability = if func.is_query() { "" } else { "mut" };
+        let name = func.name().to_case(Case::Snake);
+
+        self.code
+            .push_str(&format!("fn {}(&{} self,", name, mutability));
+        visitor::accept_func(func, self);
+        self.code.push_str(";\n");
+    }
+
+    fn visit_func_param(&mut self, func_param: &'ast FuncParam) {
+        let type_decl_code = generate_type_decl_code(func_param.type_decl());
+        self.code
+            .push_str(&format!("{}: {},", func_param.name(), type_decl_code));
+    }
+
+    fn visit_func_output(&mut self, func_output: &'ast TypeDecl) {
+        let type_decl_code = generate_type_decl_code(func_output);
+        self.code
+            .push_str(&format!(") -> Call<{}>", type_decl_code));
+    }
+}
+
+struct ClientGenerator {
+    service_name: String,
+    code: String,
+}
+
+impl ClientGenerator {
+    fn new(service_name: String) -> Self {
+        Self {
+            service_name,
+            code: String::new(),
+        }
+    }
+}
+
+impl<'ast> Visitor<'ast> for ClientGenerator {
+    fn visit_service(&mut self, service: &'ast Service) {
+        let name = &self.service_name;
+
+        self.code.push_str(&format!(
+            r#"
+            #[derive(Clone)]
+            pub struct Client<'a> {{
+                sender: &'a GStdSender
+            }}
+
+            impl<'a> Client<'a> {{
+                pub fn new(sender: &'a GStdSender) -> Self {{
+                    Self {{ sender }}
+                }}
+            }}
+
+            impl<'a> {name} for Client<'a> {{
+        "#
+        ));
+
+        visitor::accept_service(service, self);
+        self.code.push_str("}\n");
+    }
+
+    fn visit_func(&mut self, func: &'ast Func) {
+        let mutability = if func.is_query() { "" } else { "mut" };
+        let fn_name = func.name().to_case(Case::Snake);
+
+        self.code
+            .push_str(&format!("fn {}(&{} self,", fn_name, mutability));
+
+        visitor::accept_func(func, self);
+
+        self.code.push_str("{\n");
+
+        let method_name = func.name().to_case(Case::Pascal);
```

Comment:
There is no need to do any conversion here as func name comes exactly in the expected case